pkg/datapath/tables/node_address.go

```diff
@@ -0,0 +1,365 @@
+)
+
+type AddressScopeMax uint8
+
+func newAddressScopeMax(cfg NodeAddressConfig, daemonCfg *option.DaemonConfig) (AddressScopeMax, error) {
+	return AddressScopeMax(daemonCfg.AddressScopeMax), nil
+}
+
+func (cfg NodeAddressConfig) getNets() []*net.IPNet {
+	nets := make([]*net.IPNet, len(cfg.NodePortAddresses))
+	for i, cidr := range cfg.NodePortAddresses {
+		nets[i] = cidr.IPNet
+	}
+	return nets
+}
+
+func (NodeAddressConfig) Flags(flags *pflag.FlagSet) {
+	flags.StringSlice(
+		"nodeport-addresses",
+		nil,
+		"A whitelist of CIDRs to limit which IPs are used for NodePort. If not set, primary IPv4 and/or IPv6 address of each native device is used.")
+}
+
+type nodeAddressControllerParams struct {
+	cell.In
+
+	HealthScope     cell.Scope
+	Log             logrus.FieldLogger
+	Config          NodeAddressConfig
+	Lifecycle       hive.Lifecycle
+	Jobs            job.Registry
+	DB              *statedb.DB
+	Devices         statedb.Table[*Device]
+	NodeAddresses   statedb.RWTable[NodeAddress]
+	AddressScopeMax AddressScopeMax
+}
+
+type nodeAddressController struct {
+	nodeAddressControllerParams
+
+	tracker *statedb.DeleteTracker[*Device]
+}
+
+// newNodeAddressTable constructs the node address controller & registers its
+// lifecycle hooks and then provides Table[NodeAddress] to the application.
+// This enforces proper ordering, e.g. controller is started before anything
+// that depends on Table[NodeAddress] and allows it to populate it before
+// it is accessed.
+func newNodeAddressTable(p nodeAddressControllerParams) (tbl statedb.Table[NodeAddress], err error) {
+	n := nodeAddressController{nodeAddressControllerParams: p}
+	n.register()
+	return n.NodeAddresses, nil
+}
+
+func (n *nodeAddressController) register() {
+	g := n.Jobs.NewGroup(n.HealthScope)
+	g.Add(job.OneShot("node-address-update", n.run))
+
+	n.Lifecycle.Append(
+		hive.Hook{
+			OnStart: func(ctx hive.HookContext) error {
+				txn := n.DB.WriteTxn(n.NodeAddresses, n.Devices /* for delete tracker */)
+				defer txn.Abort()
+
+				// Start tracking deletions of devices.
+				var err error
+				n.tracker, err = n.Devices.DeleteTracker(txn, "node-addresses")
+				if err != nil {
+					return fmt.Errorf("DeleteTracker: %w", err)
+				}
+
+				// Do an immediate update to populate the table before it is read from.
+				devices, _ := n.Devices.All(txn)
+				for dev, _, ok := devices.Next(); ok; dev, _, ok = devices.Next() {
+					n.update(txn, nil, n.getAddressesFromDevice(dev), nil)
+				}
+				txn.Commit()
+
+				// Start the job in the background to incremental refresh
+				// the node addresses.
+				return g.Start(ctx)
+			},
+			OnStop: g.Stop,
+		})
+
+}
+
+func (n *nodeAddressController) run(ctx context.Context, reporter cell.HealthReporter) error {
+	defer n.tracker.Close()
+
+	limiter := rate.NewLimiter(nodeAddressControllerMinInterval, 1)
+	revision := statedb.Revision(0)
+	for {
+		txn := n.DB.WriteTxn(n.NodeAddresses)
```

Comment:
WriteTxn on devices is required only when adding the delete tracker (it's added to the `tableEntry` in the radix tree of tables, hence we need to lock the table to modify it). It's not needed when reading since consuming deleted objects only requires an atomic set of the revision watermark on the specific delete tracker (this watermark is read by a background garbage collector).

If you're interested in details check `(*txn).addDeleteTracker` and `pkg/statedb/graveyard.go` (garbage collector is here).